### PR TITLE
Added support for recipes conflict

### PR DESCRIPTION
**Pull request by HeahDude**
Allows to ignore incompatible recipes, helpful to avoid breaking apps using FrameworkBundle LTS versions.

flex_conflict

Unlocks https://github.com/symfony/recipes/pull/506.

We need to update the Symfony bot to allow the `conflict` key in the `manifest.json`, ping @fabpot :).

### CODE REVIEW

File: src/Flex.php [hunk 1]
```diff
@@ -298,7 +298,17 @@ public function update(Event $event, $operations = [])
             copy($rootDir.'/.env.dist', $rootDir.'/.env');
         }
 
-        $recipes = $this->fetchRecipes();
+        list($recipes, $conflicts) = $this->fetchRecipes();
+        if ($conflicts) {
+            $this->io->writeError(sprintf('<info>%d Conflict%s:</>', \count($conflicts), \count($conflicts) > 1 ? 's' : ''));
```
Comment (nicolas-grekas) [inline src/Flex.php]: I don't think any of the log messages are needed in this PR.
Which means all the logic can be moved inside fetchRecipes.
If we really want some log messages, it could be inside that method, in verbose mode only.

File: src/Flex.php [hunk 2]
```diff
@@ -594,6 +605,17 @@ private function fetchRecipes(): array
...
 
             if (isset($manifests[$name])) {
+                if (isset($manifests[$name]['manifest']['conflict'])) {
+                    foreach ($manifests[$name]['manifest']['conflict'] as $conflictingPackageName => $constraint) {
+                        if ($conflictingPackage = $this->composer->getLocker()->getLockedRepository()->findPackage($conflictingPackageName, $constraint)) {
+                            $conflicts[$name][$conflictingPackageName] = [$conflictingPackage->getPrettyVersion(), $constraint];
+                        }
+                    }
+                    unset($manifests[$name]);
+                    $this->lock->remove($name);
```
Comment (nicolas-grekas) [inline src/Flex.php]: really? feels unrelated to me.

File: src/Flex.php [hunk 3]
```diff
@@ -594,6 +605,17 @@ private function fetchRecipes(): array
             }
 
             if (isset($manifests[$name])) {
+                if (isset($manifests[$name]['manifest']['conflict'])) {
+                    foreach ($manifests[$name]['manifest']['conflict'] as $conflictingPackageName => $constraint) {
+                        if ($conflictingPackage = $this->composer->getLocker()->getLockedRepository()->findPackage($conflictingPackageName, $constraint)) {
+                            $conflicts[$name][$conflictingPackageName] = [$conflictingPackage->getPrettyVersion(), $constraint];
+                        }
+                    }
+                    unset($manifests[$name]);
```
Comment (nicolas-grekas) [inline src/Flex.php]: not needed.

File: src/Flex.php [hunk 4]
```diff
@@ -594,6 +605,17 @@ private function fetchRecipes(): array
...
+                    $this->lock->remove($name);
+
+                    continue;
```
Comment (nicolas-grekas) [inline src/Flex.php]: why is it unconditional? shouldn't it be only when a conflict has been found?

File: src/Flex.php (hunk 2 above)
Comment (HeahDude) [reply src/Flex.php]: If there is a conflict we need to revert that call, https://github.com/symfony/flex/pull/464/files#diff-46b78198ae7ea525f04268205dd782c3R602.

File: src/Flex.php (hunk 4 above)
Comment (HeahDude) [reply src/Flex.php]: >shouldn't it be only when a conflict has been found?

It is already https://github.com/symfony/flex/pull/464/files#diff-46b78198ae7ea525f04268205dd782c3R608, no?

File: src/Flex.php [hunk 5]
```diff
@@ -594,6 +594,18 @@ private function fetchRecipes(): array
             }
 
             if (isset($manifests[$name])) {
+                if (!empty($manifests[$name]['manifest']['conflict'])) {
+                    $lockedRepository = $this->composer->getLocker()->getLockedRepository();
+
+                    foreach ($manifests[$name]['manifest']['conflict'] as $conflictingPackageName => $constraint) {
+                        if ($conflictingPackage = $lockedRepository->findPackage($conflictingPackageName, $constraint)) {
+                            $this->io->writeError(sprintf('    <warning>Recipe for %s conflicts with %s:%s which should not be %s</>', $name, $conflictingPackageName, $conflictingPackage->getPrettyVersion(), $constraint));
+                            $this->lock->remove($name);
```
Comment (nicolas-grekas) [inline src/Flex.php]: A conflict should not change the state of the lock file.

File: src/Flex.php [hunk 6]
```diff
@@ -594,6 +594,18 @@ private function fetchRecipes(): array
             }
 
             if (isset($manifests[$name])) {
+                if (!empty($manifests[$name]['manifest']['conflict'])) {
+                    $lockedRepository = $this->composer->getLocker()->getLockedRepository();
+
+                    foreach ($manifests[$name]['manifest']['conflict'] as $conflictingPackageName => $constraint) {
+                        if ($conflictingPackage = $lockedRepository->findPackage($conflictingPackageName, $constraint)) {
+                            $this->io->writeError(sprintf('    <warning>Recipe for %s conflicts with %s:%s which should not be %s</>', $name, $conflictingPackageName, $conflictingPackage->getPrettyVersion(), $constraint));
```
Comment (nicolas-grekas) [inline src/Flex.php]: That's now a warning - all is fine, keep cool :)
I'd just remove the call personally. Nobody expects a recipe to come. Things work reverse: you install and you see what *might* come, without knowing beforehand.

File: src/Flex.php (hunk 2 above)
Comment (HeahDude) [reply src/Flex.php]: Got ya, fixed, thx

File: src/Flex.php (hunk 4 above)
Comment (HeahDude) [reply src/Flex.php]: Got ya, fixed, thx

File: src/Flex.php (hunk 1 above)
Comment (HeahDude) [reply src/Flex.php]: Done, without warning, on verbose

File: src/Flex.php (hunk 6 above)
Comment (HeahDude) [reply src/Flex.php]: If verbose, without warning, ok for you?

File: src/Flex.php (hunk 5 above)
Comment (HeahDude) [reply src/Flex.php]: I guess that a package should not be added to the lock if it's not configured because of a conflict

File: src/Flex.php [hunk 7]
```diff
@@ -587,6 +587,20 @@ private function fetchRecipes(): array
             $name = $package->getNames()[0];
             $job = $operation->getJobType();
 
+            if (!empty($manifests[$name]['manifest']['conflict'])) {
+                $lockedRepository = $this->composer->getLocker()->getLockedRepository();
+
+                foreach ($manifests[$name]['manifest']['conflict'] as $conflictingPackageName => $constraint) {
+                    if ($conflictingPackage = $lockedRepository->findPackage($conflictingPackageName, $constraint)) {
+                        if ($this->io->isVerbose()) {
+                            $this->io->writeError(sprintf('    Recipe for %s conflicts with %s:%s which should not be %s', $name, $conflictingPackageName, $conflictingPackage->getPrettyVersion(), $constraint));
```
Comment (nicolas-grekas) [inline src/Flex.php]: Skipping recipe for %s: it conflicts with %s %s.

Comment (HeahDude) [reply src/Flex.php]: Done.

File: src/Flex.php [hunk 8]
```diff
@@ -587,6 +587,20 @@ private function fetchRecipes(): array
             $name = $package->getNames()[0];
             $job = $operation->getJobType();
 
+            if (!empty($manifests[$name]['manifest']['conflict'])) {
+                $lockedRepository = $this->composer->getLocker()->getLockedRepository();
+
+                foreach ($manifests[$name]['manifest']['conflict'] as $conflictingPackageName => $constraint) {
+                    if ($conflictingPackage = $lockedRepository->findPackage($conflictingPackageName, $constraint)) {
+                        if ($this->io->isVerbose()) {
+                            $this->io->writeError(sprintf('  - Skipping recipe for %s: it conflicts with %s %s.', $name, $conflictingPackageName, $constraint));
```
Comment (weaverryan) [inline src/Flex.php]: What does the latest version of the message look like? I'm not sure exactly what each variable holds.

Also, does this only show in verbose mode? Does that mean it will not sure unless the user adds `-v`?

Comment (weaverryan) [reply src/Flex.php]: Ha! Ok, I feel the opposite :p. Well, sort of. Even if you assume that you don't know anything about recipes, a user expects that when they install a package, it will work just like if they follow the docs. In the case of Messenger, it means that the latest Messenger is being installed (so, ideally the user is following the latest messenger docs), but they do NO get the recipe that the docs expect. I don't see a lot of harm in showing it always - it's not a message that they will see over and over again.

I'd also like to add:
> If you want to install the recipe, fix the conflicting package(s) and run composer sync-recipes

Comment (nicolas-grekas) [reply src/Flex.php]: One wouldn't expect that in the use case that triggered the PR: if you install messenger with framework bundle, you *do not* expect it to be configured (because that's a feature of fwb >=4.1, which you decide not to use.)

Fixing a dep to get a recipe is really the reverse way of thinking: you first require deps, then you configure them, not the opposite.

Comment (weaverryan) [reply src/Flex.php]: Ok, good for me. I forgot that, for normal users, it's not possible to get a "mix" of symfony/* versions anyways, due to the locking on a specific minor thanks to Flex.

**Comment by HeahDude** [comment]
Last comments adresses, thanks for the review guys.

@weaverryan I've updated the screenshot in the description.

**Comment by fabpot** [comment]
Thank you @HeahDude.
